pkg/services/login/authinfoservice/service.go

```diff
@@ -0,0 +1,211 @@
+	"github.com/grafana/grafana/pkg/services/sqlstore"
+)
+
+const genericOAuthModule = "oauth_generic_oauth"
+
+func init() {
+	srv := &Implementation{}
+
+	registry.Register(&registry.Descriptor{
+		Name:         "UserAuthInfo",
+		Instance:     srv,
+		InitPriority: registry.MediumHigh,
+	})
+}
+
+type Implementation struct {
+	Bus                   bus.Bus                     `inject:""`
+	SQLStore              *sqlstore.SQLStore          `inject:""`
+	UserProtectionService login.UserProtectionService `inject:""`
+
+	logger log.Logger
+}
+
+func (s *Implementation) Init() error {
+	s.logger = log.New("login.authinfo")
+	return nil
+}
+
+func (s *Implementation) getUserById(id int64) (bool, *models.User, error) {
+	var (
+		has bool
+		err error
+	)
+	user := &models.User{}
+	err = s.SQLStore.WithDbSession(context.Background(), func(sess *sqlstore.DBSession) error {
+		has, err = sess.ID(id).Get(user)
+		return err
+	})
+	if err != nil {
+		return false, nil, err
+	}
+
+	return has, user, nil
+}
+
+func (s *Implementation) getUser(user *models.User) (has bool, err error) {
+	s.SQLStore.WithDbSession(context.Background(), func(sess *sqlstore.DBSession) error {
+		has, err = sess.Get(user)
+		return err
+	})
+	return
+}
+
+func (s *Implementation) LookupAndFix(query *models.GetUserByAuthInfoQuery) (bool, *models.User, *models.UserAuth, error) {
+	authQuery := &models.GetAuthInfoQuery{}
+
+	// Try to find the user by auth module and id first
+	if query.AuthModule != "" && query.AuthId != "" {
+		authQuery.AuthModule = query.AuthModule
+		authQuery.AuthId = query.AuthId
+
+		err := s.Bus.Dispatch(authQuery)
+		if !errors.Is(err, models.ErrUserNotFound) {
+			if err != nil {
+				return false, nil, nil, err
+			}
+
+			// if user id was specified and doesn't match the user_auth entry, remove it
+			if query.UserId != 0 && query.UserId != authQuery.Result.UserId {
+				err = s.Bus.Dispatch(&models.DeleteAuthInfoCommand{
+					UserAuth: authQuery.Result,
+				})
+				if err != nil {
+					s.logger.Error("Error removing user_auth entry", "error", err)
+				}
+
+				return false, nil, nil, models.ErrUserNotFound
+			} else {
+				has, user, err := s.getUserById(authQuery.Result.UserId)
+				if err != nil {
+					return false, nil, nil, err
+				}
+
+				if !has {
+					// if the user has been deleted then remove the entry
+					err = bus.Dispatch(&models.DeleteAuthInfoCommand{
+						UserAuth: authQuery.Result,
+					})
+					if err != nil {
+						s.logger.Error("Error removing user_auth entry", "error", err)
+					}
+
+					return false, nil, nil, models.ErrUserNotFound
+				}
+
+				return true, user, authQuery.Result, nil
+			}
+		}
+	}
+
+	return false, nil, nil, models.ErrUserNotFound
+}
+
+func (s *Implementation) LookupByOneOf(userId int64, email string, login string) (bool, *models.User, error) {
+	foundUser := false
+	var user *models.User
+	var err error
+
+	// If not found, try to find the user by id
+	if !foundUser && userId != 0 {
```

Comment:
```suggestion
	if userId != 0 {
```

If I'm not wrong this is always true (value equals to false) 🤔